### PR TITLE
Add Alternate Current redstone implementation (and make Eigencraft default)

### LINKED ISSUE #6975: Consider making Eigencraft Redstone default?
### Is your feature request related to a problem?

The Eigencraft Redstone engine is faster than the stock Redstone implementation, but is not default because it may cause edge incompatibilities with Redstone (though a quick questioning of devs on Discord didn't remember anything in particular). It is most likely that any incompatibilities are strictly for extremely large or technical Redstone builds, and that the vast majority of standard Redstone builds would be unaffected.

### Describe the solution you'd like.

I would like to give Eigencraft a second-look for compatibility, and if no issues are found, I don't see any reason why Eigencraft couldn't be made the default Redstone engine. In my mind, we already patch out dupe glitches but give a way out for anyone personally affected; similarly, the Eigencraft Redstone engine would just be a significant performance boost with an opt-out for anyone doing something really niche (like technical builds).

Basically, in `paper.yml`, change the default from:
`    use-faster-eigencraft-redstone: false`
to
`    use-faster-eigencraft-redstone: true`

### Describe alternatives you've considered.

I don't think this feature request has alternatives.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Titaniumtown**
(Closes https://github.com/PaperMC/Paper/issues/6975)

This PR does 3 main things:
1. (As the title describes) brings SpaceWalker's fantastic redstone implementation. [Alternate Current](https://github.com/SpaceWalkerRS/alternate-current) to Paper.
2. Makes Eigencraft the default redstone algorithm.
3. Adds a new config option in place of `use-faster-eigencraft-redstone`: `redstone-algo`

TODO:
- [X] Migrate previous `use-faster-eigencraft-redstone` config to the new `redstone-algo` option.
- [X] Use enums for config option
- [X] Properly implement config file changes in their respective patches
- [X] PaperDocs PR (https://github.com/PaperMC/PaperDocs/pull/146)

### CODE REVIEW

File: patches/server/0843-Add-Alternate-Current-redstone-implementation.patch [hunk 1]
```diff
@@ -0,0 +1,2688 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Simon Gardling <titaniumtown@gmail.com>
```
Comment (NoahvdAa) [inline patches/server/0843-Add-Alternate-Current-redstone-implementation.patch]: IMO SpaceWalkerRS should be credited as the patch author, with you as the co-author, similar to how egg did it [when porting eigencraft](https://github.com/PaperMC/Paper/pull/1494/files#diff-67d9f633149ac5490b2de8cd95b713ef29bd2749a41e022a40a23babcefe13a1R2-R7) (further discussion [here](https://github.com/PaperMC/Paper/pull/1494#discussion_r221082930))

Comment (Titaniumtown) [reply patches/server/0843-Add-Alternate-Current-redstone-implementation.patch]: Good point! I'll do that after I finished addressing some concerns that SpaceWalker brought up. Thanks!

**Comment by electronicboy** [comment]
random patches shouldn't change the configuration system used for later patches, also not entirely sure about changing the default, and not exactly fond of magic values either

**Comment by Titaniumtown** [comment]
> random patches shouldn't change the configuration system used for later patches, also not entirely sure about changing the default, and not exactly fond of magic values either

Ok, how should I go about this then? Should I merge the Eigencraft patch and this one together? and how should I ideally manage the config option?

Edit: SpaceWalker suggested using enums, I'll do that in a bit.

File: patches/server/0843-Add-Alternate-Current-redstone-implementation.patch (hunk 1 above)
Comment (Titaniumtown) [reply patches/server/0843-Add-Alternate-Current-redstone-implementation.patch]: Fixed!

**Comment by Titaniumtown** [comment]
I've now further improved the config option handling. It now uses a string that's converted into an enum and isn't just a "magic value" (as electronicboy put it) xD. It also now should migrate the old eigencraft config option!

**Comment by Titaniumtown** [comment]
The config changes are now implemented properly. And all the new added classes are under `io.papermc.paper.alternatecurrent`. I'll work on a PaperDocs PR to coincide with this PR.

Edit: the original author of Alternate Current, SpaceWalker, said they preferred if Alternate Current's classes were in their own group. So they are now located in `alternate.current`.

**Comment by justadudeongithub** [comment]
@Titaniumtown thank you for implementing this much awaited patch. Why was the merge blocked?

**Comment by Titaniumtown** [comment]
> @Titaniumtown thank you for implementing this much awaited patch. Why was the merge blocked?

I do not know.

**Comment by Owen1212055** [comment]
See: https://github.com/PaperMC/Paper/issues/6975#issuecomment-1056708772

Basically Eigencraft shouldn't be default.

**Comment by LaurenceBarnes** [comment]
> See: [#6975 (comment)](https://github.com/PaperMC/Paper/issues/6975#issuecomment-1056708772)
> 
> Basically Eigencraft shouldn't be default.

Is there anything else preventing this from being merged? @Titaniumtown looks like the docs PR has to be redone since the changes happened

**Comment by electronicboy** [comment]
needs testing, analysis, etc, etc

**Comment by dkbay** [comment]
Why was this closed? 👀 
I was really looking forward to this.

**Comment by Titaniumtown** [comment]
I'm unmotivated to work on this any longer (and minecraft as a whole), I've moved on.

**Pull request closed without merging**